archive/tar_stream.ts

```diff
@@ -0,0 +1,466 @@
+// Copyright 2018-2024 the Deno authors. All rights reserved. MIT license.
+
+/**
+ * The interface required to provide a file.
+ */
+export interface TarStreamFile {
+  /**
+   * The name of the file.
+   */
+  pathname: string;
+  /**
+   * The size of the file.
+   */
+  size: number;
+  /**
+   * The contents of the file.
+   */
+  readable: ReadableStream<Uint8Array>;
+  /**
+   * The metadata of the file.
+   */
+  options?: TarStreamOptions;
+}
+
+/**
+ * The interface required to provide a directory.
+ */
+export interface TarStreamDir {
+  /**
+   * The name of the directory.
+   */
+  pathname: string;
```

Comment:
@BlackAsLight I see you dismissed these suggestions, but made similar changes in recents commits. It might be useful to communicate the additional constraints for these values: that the paths need to be posix-style paths (separators are `/` characters), and that directories need to end with a separator. Some users will be coming from Windows backgrounds and might think that it will work to use a relative, native filesystem path.

Reply:
I marked them as resolved as I made the parts I thought relevant in the `doc(archive): updated to be more descriptive` commit.

While it is customary for directories to end with `/` for tar implementations, it isn't a requirement of the tar spec. 

As for the path, it can contain any character as the tar spec only specifies a byte length limit. If it is greater than 100 then it can only be split on a `/`. Tar allows paths to contain characters that would normally be an invalid path on an OS.